sdk_extensions/zpages/src/main/java/io/opentelemetry/sdk/extensions/zpages/TracezSpanProcessor.java

```diff
@@ -0,0 +1,215 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.extensions.zpages;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.opentelemetry.sdk.common.export.ConfigBuilder;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+import io.opentelemetry.sdk.trace.SpanProcessor;
+import io.opentelemetry.trace.SpanId;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A {@link SpanProcessor} implementation for the traceZ zPage.
+ *
+ * <p>Configuration options for {@link TracezSpanProcessor} can be read from system properties,
+ * environment variables, or {@link java.util.Properties} objects.
+ *
+ * <p>For system properties and {@link java.util.Properties} objects, {@link TracezSpanProcessor}
+ * will look for the following names:
+ *
+ * <ul>
+ *   <li>{@code otel.zpages.export.sampled}: sets whether only sampled spans should be exported.
+ * </ul>
+ *
+ * <p>For environment variables, {@link TracezSpanProcessor} will look for the following names:
+ *
+ * <ul>
+ *   <li>{@code OTEL_ZPAGES_EXPORT_SAMPLED}: sets whether only sampled spans should be exported.
+ * </ul>
+ */
+@ThreadSafe
+final class TracezSpanProcessor implements SpanProcessor {
+  private final ConcurrentMap<SpanId, ReadableSpan> runningSpanCache;
+  private final ConcurrentMap<String, TracezSpanBuckets> completedSpanCache;
+  private final Set<String> spanNames;
+  private final boolean sampled;
+
+  /**
+   * Constructor for {@link TracezSpanProcessor}.
+   *
+   * @param sampled report only sampled spans.
+   */
+  public TracezSpanProcessor(boolean sampled) {
+    runningSpanCache = new ConcurrentHashMap<>();
+    completedSpanCache = new ConcurrentHashMap<>();
+    spanNames = new TreeSet<>();
+    this.sampled = sampled;
+  }
+
+  @Override
+  public void onStart(ReadableSpan span) {
+    runningSpanCache.putIfAbsent(span.getSpanContext().getSpanId(), span);
+    synchronized (spanNames) {
+      spanNames.add(span.getName());
+    }
+  }
+
+  @Override
+  public boolean isStartRequired() {
+    return true;
+  }
+
+  @Override
+  public void onEnd(ReadableSpan span) {
+    runningSpanCache.remove(span.getSpanContext().getSpanId());
+    if (!sampled || span.getSpanContext().getTraceFlags().isSampled()) {
+      completedSpanCache.putIfAbsent(span.getName(), new TracezSpanBuckets());
```

Comment:
Ditto, we'll want to do a double-checked lock

```java
TracezSpanBuckets buckets = completedSpanCache.get(span.getName());
if (buckets == null) {
  synchronized (completedSpanCache) {
    buckets = completedSpanCache.get(span.getName());
    if (buckets == null) {
      buckets = new TracezSpanBuckets();
      completedSpanCache.put(span.getName(), buckets);
    }
  }
}
buckets.addToBucket(span);
```

Reply:
Is this still the preferable implementation when `completedSpanCache` is a `ConcurrentMap`?

Reply:
Honestly, I'm not sure it matters all that much. The worst case scenario would be that we miss a couple of spans at startup time. I'd prefer to keep the code simpler, since this is a non-critical code path (temporary losses are totally fine, since we're only keeping a few spans anyway).